src/viur/core/skeleton.py

```diff
@@ -1141,9 +1180,9 @@ def __txn_update(write_skel):
 
             db_obj = skel.preProcessSerializedData(db_obj)
 
-            # Allow the custom DB Adapter to apply last minute changes to the object
-            if skel.customDatabaseAdapter:
-                db_obj = skel.customDatabaseAdapter.preprocessEntry(db_obj, skel, change_list, is_add)
+            # Allow the database adapter to apply last minute changes to the object
+            for adapter in skel.database_adapters:
+                adapter.prewrite(skel, is_add, change_list)
```

Comment:
**First;** The object can be changed in all situations.
**Second;** No, the *db_obj* is contained in *skel* and can be accessed by *skel.dbEntity*. It can freely be changed.
**Third;** No, its bad naming convention and should be eliminated. I want to have some kind of hooks named pre-<action>, <action>, post-<action>. This is a concept. And a pre-hook generally is a hook and nothing which returns any object, which *might* be changed. It still can be. Therefore, I want to keep the concept now as it is, especially because it is more concept with conciseness than the messed-up ugly stuff below, left untended and unloved for a decade.

Reply:
Additionally, to underline the reasons for this naming convention, is that #630 already defines a similar concept, although much is missing. The current situation is a mess, and not a standard.

Reply:
> The object can be changed in all situations

Yes, but it should obviously be recognisable that this method changes an object. And you make this recognisable via a return value and not just via side effects.



> No, the _db_obj_ is contained in _skel_ and can be accessed by _skel.dbEntity_. It can freely be changed

If my custom `preProcessSerializedData` creates a new instance, you would completely ignore this object in this method, but this new object would be written later with `db.Put`.

```
def preProcessSerializedData(self, db_obj):
    new_obj = db.Entity()
    # [...] Some edge case copy logic or whatever
    return new_obj
```

Therefore, you need to continue with the `db_obj` and not with `skel.dbEntity`, they must no longer be the identical one.





> No, its bad naming convention and should be eliminated. I want to have some kind of hooks named _pre-action_, _action_, and _post-action_, although _action_ is the post-action. This is a concept

Well, it's just a name. Call it what you like. I still think it fits better under a known name, considering the existing code. And then later in ViUR4 it's better to rename everything consistently than to call it _preprocess_ in one class and _prewrite_ in another.




> And a pre-hook generally is a hook and nothing which returns any object, which _might_ be changed.

I don't know what makes you think that a hook doesn't return anything, but a hook can very well change objects, return them and thus change the behaviour of software. That's exactly the point of hooks, to create the ability to intervene in code. There are several examples on [Wikipedia](https://en.wikipedia.org/wiki/Hooking) and in the linked literature.

Reply:
@sveneberth I've now updated the pull request again.

> Therefore, you need to continue with the `db_obj` and not with `skel.dbEntity`, they must no longer be the identical one.

Well, there should be no further distinction between the Skeleton instance and the database object that is written. This is a datastore-specific thing, and should not be part of any hooks, the hook should work with the skeleton instance object and its members, and not with the skeleton instance object and a copy of the content as db.Entity.

It is nonsense to return and store and entirely different database object than the `skel.dbEntity`-one. To get rid of the `db_obj` variable handled inside the transactional function, I've now replaced it so that `skel.dbEntity` is used throughout the entire process.

You can still replace it by everything you want, on your demand and on your responsibility. The name `prewrite` is more clerly than "preprocess"... preprocesses what? C source code, like a C-pre-processor #ifdef #ifndef, or what? (scnr).

This is my last comment on this issue now. I need the feature for an important and critical project and also a cleaner code base, which is not the result of several years of omitted maintenance.